Add Erlang-inspired process group (`pg`) scopes

leonqadirie opened the pull request:
Might resolve #31.

I went the proposed route with split functions and `(scope, group)` as an identifying tuple.
Couldn't find the mentioned protobuf OTW cluster protocol changes.

As this is my first-ever PR with 'real' code I suggest a more thorough look. My main concerns besides the API design are:
* correct handling of monitors, and adjacently,
* whether I should handle idempotency of received messages if an `ActorCell` is part of multiple scopes and groups. Couldn't find any mechanisms for the existing process group implementation, so I assume this is not necessary.

Thanks!

--
Edit: Regarding the code coverage: this seems to mostly stem from duplicated yet also previously untested logic due to the split function approach. The exception is the new `which_scopes_and_groups()` for which I can provide a test if desired - just a bit trickier when running tests concurrently. :)

slawlor commented:
Thanks for the contribution! I'll try and take a look as soon as I can but it might be a couple of days. 😁

slawlor commented on ractor/src/pg/mod.rs:
```diff
@@ -206,14 +351,57 @@ pub fn get_local_members(group_name: &GroupName) -> Vec<ActorCell> {
+/// * `group_name` - Either a statically named group
+///
+/// Returns a [`Vec<ActorCell>`] representing the members of this paging group
+pub fn get_local_members_with_named_scope(
```
nit: same naming and the non-scoped call could call the scoped version

slawlor commented on ractor/src/pg/mod.rs:
```diff
@@ -206,14 +351,57 @@ pub fn get_local_members(group_name: &GroupName) -> Vec<ActorCell> {
+/// * `group_name` - Either a statically named group or scope
+///
+/// Returns a [`Vec<ActorCell>`] with the member actors
+pub fn get_members_with_named_scope(scope: &ScopeName, group_name: &GroupName) -> Vec<ActorCell> {
```
nit: same naming and call the non-scoped version into this

slawlor commented on ractor/src/pg/mod.rs:
```diff
@@ -224,6 +412,37 @@ pub fn get_members(group_name: &GroupName) -> Vec<ActorCell> {
+        .map
+        .iter()
+        .map(|kvp| kvp.key().clone())
+        .map(|(_scope, group)| group.clone())
```
groups can be in multiple scopes I think, we probably need to de-duplicate this list

slawlor commented on ractor/src/pg/mod.rs:
```diff
@@ -224,6 +412,37 @@ pub fn get_members(group_name: &GroupName) -> Vec<ActorCell> {
+    monitor
+        .map
+        .iter()
+        .map(|kvp| kvp.key().clone())
```
This is where having another index might be helpful, i.e. Scope -> Group name in a separate mapping. But that's an optimization that can be added in the future.

leonqadirie replied:
Left for the future :)

slawlor reviewed (changes requested):
I think we can reduce a lot of code duplication by having the non-scoped methods call the scoped ones just with the default scope as an argument. That also will reduce it to 1 implementation of each call (join/leave/etc)

leonqadirie commented:
> I think we can reduce a lot of code duplication by having the non-scoped methods call the scoped ones just with the default scope as an argument. That also will reduce it to 1 implementation of each call (join/leave/etc)

Well, duh. Thanks for the pointers, learned a bit or two.
I did get occasional CI failures for the async-std tests, but suspect they are not related to my code.

slawlor commented:
> I did get occasional CI failures for the async-std tests, but suspect they are not related to my code.

Yeah I'm working on stabilizing these actually. I have another branch I'm going to split out some changes for this runtime to stabilize the flaky tests. I'm working on a wasm runtime but that might be a dead end lol

slawlor commented:
Apologies but this might take some time to get to another review. We just had a new baby and I'm a little sleep deprived at the moment lol

leonqadirie commented:
First of all, congratulations! And no worries, take all the time you need. :)

slawlor commented:
I think it's really close! but there's some cleanups and optimizations that could be done. Should be quick & minimal. Thanks again!

leonqadirie commented:
Cool and thanks for stabilizing the async-std tests!

A bit unsure of how to proceed/which cleanups and optimizations you are referring to specifically.
Besides possibly this review, that is:
> This is where having another index might be helpful, i.e. Scope -> Group name in a separate mapping. But that's an optimization that can be added in the future.

slawlor commented:
I think I missed a button to.mark.the review done. I'm happy to leave the double index but to the future but the 6 comments in my latest review should be addressed prior to merging. If anything doesn't make sense, let me know. And feel free to mark the old review comments as done to clear up the pending comments

leonqadirie commented on ractor/src/pg/tests.rs:
```diff
@@ -52,6 +52,66 @@ async fn test_basic_group() {
+    handle.await.expect("Actor cleanup failed");
+}
+
+// #[named]
```
Ah I didn't unstage that one, sorry.

As far as I understand it, my challenge with this test is that it runs concurrently with the other tests while asserting on the monitor's state. This results in unreliability.

The following tests specifically seem to interfere with this one:
* `test_pg_monitoring()`
* `test_scope_monitoring()`

If we want to test this functionality (which would be my preference), my current ideas are to
* either limit test concurrency to start the other tests after this one is done, or to
* filter any `ScopeGroupKey`s from the other tests explicitly from assertion before asserting (increases this test's projected maintenance effort), or to
* only asserting the expected `ScopeGroupKey`s are in the global state (reduced projected maintenance effort, but wouldn't check whether within the test we accidentally did create unexpected `ScopeGroupKey`s).

Happy to hear your thoughts :)

slawlor replied:
We can followup on this one in another PR, but basically you can serialize the tests with [`serial_test`](https://docs.rs/serial_test/latest/serial_test/) which asserts that only one test can run at a time (without messing with other concurrency settings, etc).

Additionally I'm happy to work with specifically named keys that wouldn't conflict, both are fine for me in this scenario.

slawlor reviewed (approved):
Thanks for addressing the comments and bearing with me! This looks great, and thanks for the additions :)

leonqadirie commented:
> Thanks for addressing the comments and bearing with me! This looks great, and thanks for the additions :)

Nice! Feels good to have my first 'real' PR merged. :)
Thanks for the feedback and please always feel free to be (constructively) pedantic with my code, I already learned a ton here.

Will address the test and then maybe take a look into what kind of index optimisation you had in mind.
